File: plugins/keyboard/ios/templates/KeyAdjacency.swift
import Foundation

/// QWERTY key-neighborhood table for the fat-finger substitution model:
/// substituting a physically adjacent key costs less than a random one.
/// Derived from the staggered 3-row layout (horizontal neighbors plus the
/// nearest keys in the adjacent rows).
///
/// Must stay identical to `KEY_ADJACENCY` in
/// `scripts/keyboard-dictionary/decoder.js` (the reference implementation)
/// and `KeyAdjacency.kt` (the Android twin).
enum KeyAdjacency {

    private static let neighbors: [Character: String] = [
        "q": "wa", "w": "qeas", "e": "wrsd", "r": "etdf", "t": "ryfg",
        "y": "tugh", "u": "yihj", "i": "uojk", "o": "ipkl", "p": "ol",
        "a": "qwsz", "s": "weadzx", "d": "ersfxc", "f": "rtdgcv", "g": "tyfhvb",
        "h": "yugjbn", "j": "uihknm", "k": "iojlm", "l": "opk",
        "z": "asx", "x": "sdzc", "c": "dfxv", "v": "fgcb", "b": "ghvn",
        "n": "hjbm", "m": "jkn",
    ]

    /// Bitset over ASCII pairs so the DP inner loop pays a table read, not a
    /// dictionary hash. Non-letters (`'`, `-`) have no neighbors.
    private static let table: [Bool] = {
        var table = [Bool](repeating: false, count: 128 * 128)
        for (key, adjacent) in neighbors {
            let a = Int(key.asciiValue!)
            for n in adjacent.utf8 {
                table[a * 128 + Int(n)] = true
            }
        }
        return table
    }()

    static func isAdjacent(_ a: UInt8, _ b: UInt8) -> Bool {
        guard a < 128, b < 128 else { return false }
        return table[Int(a) * 128 + Int(b)]
    }

    /// Letter-key centers in key-grid units (key width = 1.0) on the standard
    /// QWERTY 10/9/7 layout — rows at y 0.5/1.5/2.5, home and bottom rows
    /// indented by 0.5 and 1.5 key widths. Mirrors `KEY_CENTERS` in decoder.js;
    /// native key views normalize taps into this same space.
    private static let centers: [(x: Float, y: Float)?] = {
        var centers = [(x: Float, y: Float)?](repeating: nil, count: 128)
        let rows: [(String, Float, Float)] = [
            ("qwertyuiop", 0.5, 0.5),
            ("asdfghjkl", 1.0, 1.5),
            ("zxcvbnm", 2.0, 2.5),
        ]
        for (letters, x0, y) in rows {
            for (i, ch) in letters.utf8.enumerated() {
                centers[Int(ch)] = (x0 + Float(i), y)
            }
        }
        return centers
    }()

    static func center(_ a: UInt8) -> (x: Float, y: Float)? {
        guard a < 128 else { return nil }
        return centers[Int(a)]
    }
}

File: plugins/keyboard/android/templates/KeyAdjacency.kt
package com.a1lab.echos.ime

/**
 * QWERTY key-neighborhood table for the fat-finger substitution model:
 * substituting a physically adjacent key costs less than a random one.
 * Derived from the staggered 3-row layout (horizontal neighbors plus the
 * nearest keys in the adjacent rows).
 *
 * Must stay identical to `KEY_ADJACENCY` in
 * `scripts/keyboard-dictionary/decoder.js` (the reference implementation)
 * and `KeyAdjacency.swift` (the iOS twin).
 */
object KeyAdjacency {

    private val neighbors = mapOf(
        'q' to "wa", 'w' to "qeas", 'e' to "wrsd", 'r' to "etdf", 't' to "ryfg",
        'y' to "tugh", 'u' to "yihj", 'i' to "uojk", 'o' to "ipkl", 'p' to "ol",
        'a' to "qwsz", 's' to "weadzx", 'd' to "ersfxc", 'f' to "rtdgcv",
        'g' to "tyfhvb", 'h' to "yugjbn", 'j' to "uihknm", 'k' to "iojlm",
        'l' to "opk",
        'z' to "asx", 'x' to "sdzc", 'c' to "dfxv", 'v' to "fgcb", 'b' to "ghvn",
        'n' to "hjbm", 'm' to "jkn",
    )

    /** Bitset over ASCII pairs so the DP inner loop pays a table read, not a
     *  map hash. Non-letters (`'`, `-`) have no neighbors. */
    private val table = BooleanArray(128 * 128).also { table ->
        for ((key, adjacent) in neighbors) {
            for (n in adjacent) {
                table[key.code * 128 + n.code] = true
            }
        }
    }

    fun isAdjacent(a: Byte, b: Byte): Boolean {
        val ai = a.toInt() and 0xFF
        val bi = b.toInt() and 0xFF
        if (ai >= 128 || bi >= 128) return false
        return table[ai * 128 + bi]
    }

    /** Letter-key centers in key-grid units (key width = 1.0) on the standard
     *  QWERTY 10/9/7 layout — rows at y 0.5/1.5/2.5, home and bottom rows
     *  indented by 0.5 and 1.5 key widths. Mirrors `KEY_CENTERS` in decoder.js;
     *  native key views normalize taps into this same space. Packed as
     *  [x0, y0, x1, y1, ...] indexed by ASCII code; NaN marks a non-letter. */
    private val centersX = FloatArray(128) { Float.NaN }
    private val centersY = FloatArray(128) { Float.NaN }

    init {
        val rows = listOf(
            Triple("qwertyuiop", 0.5f, 0.5f),
            Triple("asdfghjkl", 1.0f, 1.5f),
            Triple("zxcvbnm", 2.0f, 2.5f),
        )
        for ((letters, x0, y) in rows) {
            for ((i, ch) in letters.withIndex()) {
                centersX[ch.code] = x0 + i
                centersY[ch.code] = y
            }
        }
    }

    /** Returns the key center as (x, y) in grid units, or null for non-letters. */
    fun center(a: Byte): Pair<Float, Float>? {
        val ai = a.toInt() and 0xFF
        if (ai >= 128 || centersX[ai].isNaN()) return null
        return centersX[ai] to centersY[ai]
    }
}
